streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1094,16 +1100,32 @@ private int getNumStreamThreads(final boolean hasGlobalTopology) {
     }
 
     private int getNextThreadIndex() {
-        final HashSet<String> names = new HashSet<>();
-        processStreamThread(thread -> names.add(thread.getName()));
-        final String baseName = clientId + "-StreamThread-";
-        for (int i = 1; i <= threads.size(); i++) {
-            final String name = baseName + i;
-            if (!names.contains(name)) {
-                return i;
+        final HashSet<String> allLiveThreadNames = new HashSet<>();
+        final AtomicInteger maxThreadId = new AtomicInteger(1);
+        synchronized (threads) {
+            processStreamThread(thread -> {
+                // trim any DEAD threads from the list so we can reuse the thread.id
+                // this is only safe to do once the thread has fully completed shutdown
+                if (thread.state() == StreamThread.State.DEAD) {
+                    threads.remove(thread);
+                } else {
+                    allLiveThreadNames.add(thread.getName());
+                    final int threadId = thread.getName().charAt(thread.getName().length() - 1);
```

Comment:
If a Streams client has a number of stream threads that is not single digit (i.e. 10+), this does not work anymore.

Reply:
Oh duh. That would have been an annoying bug to find in production 😅